lib/pos_mempool_persister.go

```diff
@@ -0,0 +1,205 @@
+)
+
+type MempoolEventType int
+
+const (
+	MempoolEventAdd MempoolEventType = iota
+	MempoolEventRemove
+	MempoolEventExit
+)
+
+type MempoolPersisterStatus int
+
+const (
+	MempoolPersisterStatusRunning MempoolPersisterStatus = iota
+	MempoolPersisterStatusNotRunning
+)
+
+// MempoolPersister is responsible for persisting transactions in the mempool to the database. Whenever a transaction is
+// added or removed from the mempool, the MempoolPersister is notified via the OnMempoolEvent callback. The Persister
+// will then add the event to a queue. Periodically, the transaction queue is flushed to the database and all the cached
+// transactions are persisted.
+type MempoolPersister struct {
+	sync.Mutex
+	stopGroup  sync.WaitGroup
+	startGroup sync.WaitGroup
+
+	frequencyMilliseconds int
+	status                MempoolPersisterStatus
+	dbCtx                 *storage.DatabaseContext
+	eventQueue            chan *MempoolEvent
+	updateBatch           []*MempoolEvent
+}
+
+func NewMempoolPersister(dbCtx *storage.DatabaseContext, frequencyMilliseconds int) *MempoolPersister {
+	return &MempoolPersister{
+		frequencyMilliseconds: frequencyMilliseconds,
+		status:                MempoolPersisterStatusNotRunning,
+		dbCtx:                 dbCtx,
+		eventQueue:            make(chan *MempoolEvent, eventQueueSize),
+	}
+}
+
+func (mp *MempoolPersister) Start() {
+	if mp.status == MempoolPersisterStatusRunning {
+		return
+	}
+
+	mp.reset()
+	mp.stopGroup.Add(1)
+	mp.startGroup.Add(1)
+	go mp.run()
+	mp.startGroup.Wait()
+	mp.status = MempoolPersisterStatusRunning
+}
+
+func (mp *MempoolPersister) run() {
+	mp.startGroup.Done()
+	for {
+		select {
+		case event := <-mp.eventQueue:
+			switch event.Type {
+			case MempoolEventAdd, MempoolEventRemove:
+				mp.Lock()
+				mp.updateBatch = append(mp.updateBatch, event)
+				mp.Unlock()
+
+			case MempoolEventExit:
+				mp.status = MempoolPersisterStatusNotRunning
+				mp.stopGroup.Done()
+				return
+			}
+			continue
+
+		case <-time.After(time.Duration(mp.frequencyMilliseconds) * time.Millisecond):
+			if err := mp.persistBatch(); err != nil {
+				glog.Errorf("MempoolPersister: Error persisting batch: %v", err)
+			}
+			continue
+		}
+	}
+}
+
+func (mp *MempoolPersister) Stop() error {
+	if mp.status == MempoolPersisterStatusNotRunning {
+		return nil
+	}
+	mp.eventQueue <- &MempoolEvent{Type: MempoolEventExit}
+	mp.stopGroup.Wait()
+	if err := mp.persistBatch(); err != nil {
+		return errors.Wrapf(err, "MempoolPersister: Error persisting batch")
+	}
+	mp.reset()
+	return nil
+}
+
+func (mp *MempoolPersister) persistBatch() error {
+	if mp.status == MempoolPersisterStatusNotRunning {
+		return nil
+	}
+
+	mp.Lock()
+	defer mp.Unlock()
+
+	if len(mp.updateBatch) == 0 {
+		return nil
+	}
+
+	localContext := mp.dbCtx.GetContext([]byte(DbMempoolContextId))
+	err := mp.dbCtx.Update(localContext, func(txn storage.Transaction, ctx storage.Context) error {
+		for _, event := range mp.updateBatch {
```

Comment:
This function and the next one are the only ones that read/write to the DB. And the DB operations they make are quite simple. Set, Delete, and Iteration+Read. Performing the same operations using Badger's API would be just as straightforward. 

If that's the case, then the DB interface in the previous PR doesn't add much (any?) value is making it easier to swap in a new DB. But the interface is a quite large middle layer we need to maintain.

There's a case to be made for deprecating the DB interface, and moving it to your experimental branch. And to instead perform the DB operations here using Badger's API directly.